### ISSUE TITLE
Reindexing is disruptive in stable deployments

### ISSUE THREAD

**Issue by hannes-ucsc** [DataBiosphere/azul#6026]
Users complain that deployments are slow during reindexing, that the index is incomplete and that they can't download manifests. To address this we would need to perform a reindex in the background, so to speak, ideally without doubling the infrastructure cost.

**Comment by dsotirho-ucsc** [comment]
Assignee to consider next steps.

**Comment by hannes-ucsc** [comment]
[edit 2024/12/27 @hannes-ucsc] Incorporating #5122

We've been using the terms *deployment* and *instance* interchangeably but they'll be assigned different meanings with this work. For stable deployments (e.g. `prod` and `anvilprod`), we'll simultaeneously maintain two instances of Azul per deployment. The approach is based on [Blue-Green deployments](https://en.wikipedia.org/wiki/Blue%E2%80%93green_deployment) but we'll refer to the instances as 1 and 2, as in `prod1` and `prod2`. An instance can be either live or stand-by. Only one instance in a deployment can be live, and stand-by instances can optionally be inactive. Rarely will there be more than two instances per deployment.

Given the `prod` example, the instance environment will be configured in `deployments/prod/environment.py`, as before, but that file will contain the following section:

```
{
	…
	'azul_selected_instance': '1', # or 2
	'AZUL_DEPLOYMENT_STAGE': 'prod{azul_selected_instance}',
	'AZUL_DOMAIN_NAME': 'explore.data.humancellatlas.org',
	'AZUL_SUBDOMAIN_TEMPLATE': '*.{azul_selected_instance}',
	…
}
```

The new variable `azul_selected_instance` (lowercase, so that it will not be exported), controls which instance is targetted by `make deploy` and related operations. The TF components `prod.shared`, `prod.gitlab` and `prod.browser` will override `azul_selected_instance` with `None` (or `''`) and `AZUL_DEPLOYMENT_STAGE` with `'prod'`, thereby ensuring that there is only one instance of these components. In the future, we may consider maintaining two instances of the `browser` component.

The central piece that makes this work is the fact that the TF state file name is derived from `AZUL_DEPLOYMENT_STAGE`, so modifying `azul_selected_instance` will select a different TF state file, and therefore a different set of cloud resources. The sets will be distinct due to the fact that all resource names in a main TF component are qualified with `AZUL_DEPLOYMENT_STAGE`.

If a promotion requires a reindex or some other change causing degradation of service or outright downtime, the PR branch for that promotion will contain a commit that modifies `azul_selected_instance` from 1 to 2 or vice versa. This commit will not be backported. The `develop` version of `deployments/prod/environment.py` should not even contain an entry for `azul_selected_instance` so that there is a permanent difference in that file between the `prod` and `develop` branches. Within that file, the entry should be placed in an area that is rarely modified, so as to minimize conflicts during promotions or backports.

With the above environment, the production domain names will be

- `service.1.explore.data.humancellatlas.org`
- `indexer.1.explore.data.humancellatlas.org`

for the first instance, and

- `indexer.2.explore.data.humancellatlas.org`
- `service.2.explore.data.humancellatlas.org`

for the second. These are the canonical domain names of the indexer and service in either instance. Note that the hosted zone (as set by `AZUL_DOMAIN_NAME`) is the same for both instances.

Additionally, there will be alias A records that point to one of the instances:

- `indexer.explore.data.humancellatlas.org` -> `indexer.1.explore.data.humancellatlas.org`
- `service.explore.data.humancellatlas.org` -> `service.1.explore.data.humancellatlas.org`

if instance 1 is live or 

- `indexer.explore.data.humancellatlas.org` -> `indexer.2.explore.data.humancellatlas.org`
- `service.explore.data.humancellatlas.org` -> `service.2.explore.data.humancellatlas.org`

if instance 2 is.

Note that alias A records are different from CNAME records. The latter require an extra DNS round-trip while the former don't. Users won't typically be aware of the canonical domain names. Publicly, we'll communicate only the alias record names. According to the [AWS documentation](https://docs.aws.amazon.com/Route53/latest/DeveloperGuide/resource-record-sets-choosing-alias-non-alias.html), aliasing one A record to another one is supported as long as both records are in the same hosted zone:

> Another Route 53 record of the same type in the same hosted zone – Route 53 responds as if the query is for the record that is referenced by the alias record (see Comparison of alias and CNAME records).

Each ACM certificate in these two instances will list the canonical domain name as the common name and the alias record as the subject alternative name. This will allow the deployment to be reached at both the canonical domain name (unless it is inactive) and the alias (if it is live). We already have an aliasing mechanism for the DRS domain name. We will remove the DRS alias and retool the mechanism to only provide the certificate, the validation record in Route 53, but not the alias A record, since the latter will be maintained by a script, outside of Terraform.

Note that the canonical record is also an alias, but to a CloudFront distribution managed internally by API Gateway. I experimentally confirmed that it is possible to have an alias A record of another alias A record.

Live and stand-by instances will have dedicated ES domains. The ES domain of an inactive instance will not exist.

During a reindexing operation, the live instance will serve external users and host the current version of the code and index. The stand-by instance will host the next version of the code and will be running a reindex. When the reindex is done, the alias records will be changed to point to the stand-by instance, so that it goes live, and putting the other instance into stand-by. After verifying that the live instance is functional, the other instance will be deactivated. This will cause it's expensive resources, such as the ES domain.

Scripts will be used to manage the alias records and to activate/deactivate instances. The standard GitLab pipeline will have dedicated manual jobs for these functions, invoking the scripts. Having these jobs reduces the potential for human error (wrong deployment or instance selected, branch check out, branch stale). The `deactivate` job will deactivate the instance selected by `deployments/prod/environment.py` in the commit the pipeline is asscociated with. The `live` job will adjust the alias record to the selected instances. The already existing `deploy` job will undo the effect of the `deactivate`. We already have shell function called `hibernate` that we may want to consider adopting or consolidating with the script for deactivation.

For the addition of a catalog for, say, a new HCA release, the operator either promotes the catalog change to the `prod` branch via a promotion PR, or creates a hotfix PR with that change. To either branch the operator adds a commit that alters `azul_selected_instance` to point to one of the inactive or stand-by instances. Typically, there'll be just one such instance to choose from. Note that selecting an instance is different from switching the alias records. The operator then merges the promotion branch or hotfix branch. Pushing the merge commit to the GitLab instance will trigger a pipeline. The `deploy` job of that pipeline will activate the selected instance if necessary, putting it into stand-by. Once the `integration_test` job passes, the operator will manually start the reindex job. When the reindex is complete, the operator will run the `live` job and the newly added catalog will be observable to external clients of the Azul API. The operator will then run the `deploy_browser` job of that pipeline. This will make the catalog available to users of the Data Browser, provided that they add `?catalog=…` to the Data Browser URL. Once the wranglers approve the release, and once the default catalog in the Data Browser was updated by CC, the operator merge the Data Browser update and push the merge to GitLab. This will deposit the Data Browser tarball on the GitLab instance. The operator will then author another PR (typically targetting `develop`) that removes the catalog for the current release, thereby making the new release the default catalog for explore. Unless there is any other change requiring a reindex, the promotion PR will not need to conatain a commit that alters `azul_selected_instance`. The `deploy_browser` (`deploy`) job will affectuate the default catalog change for the Data Browser (Azul), thereby making the new release current for users of the Data Browser (Azul). Lastly, the operator will manually delete the catalog for the previous release.


TODO: Create child issues for

- Removing the DRS alias (but not the aliasing mechanism, leave the certificate alt names)
- Adding the ability to have multiple instances per deployment
- Replacing `prod` instance with `prod1` instance (probably as a hotfix)
- Adding the `prod2` instance (and switching alias to it)
- Adding `live` and `deactivate` jobs to GitLab build
- Updating PR checklists to reflect instance management
- Updating the compliance documentation

**Comment by bvizzier-ucsc** [comment]
Moving to Triage.  We need to get an effort estimate on this.

**Comment by hannes-ucsc** [comment]
I simplified the design (no need for multiple branches anymore). Still need to proof-read and create child tickets.

**Comment by hannes-ucsc** [comment]
This is still held up by upcoming assessment.

**Comment by hannes-ucsc** [comment]
Pen test is complete so we could pick this up but mirroring is getting hot so there is no time for this currently.

**Comment by hannes-ucsc** [comment]
I'd like to get this re-prioritized in light of the other epics. I'd also like PM to ensure that we have epics for all priorities the team is supposed to pursuit.

Assignee to facilitate these two requests.
